tests/factories.py

```diff
@@ -0,0 +1,92 @@
+import base64
+import secrets
+
+from jobrunner import job_executor
+from jobrunner.models import Job, State
+from jobrunner.lib.database import insert
+from jobrunner.manage_jobs import JobError
+
+
+DEFAULTS = {
+    "action": "action_name",
+    "repo_url": "opensafely/study",
+    "workspace": "workspace",
+    "requires_outputs_from": [],
+    "run_command": "python myscript.py",
+    "output_spec": {},
+}
+
+
+def job_factory(**kwargs):
+    values = DEFAULTS.copy()
+    values.update(kwargs)
+    if "job_request_id" not in values:
+        # use random job_request_id as id is derived from it
+        values["job_request_id"] = base64.b32encode(
+            secrets.token_bytes(10)
+        ).decode("ascii").lower()
+    return Job(**values)
+
+
+class TestJobAPI():
+
+    def __init__(self):
+        self.jobs = {}
+        self.jobs_run = {}
+        self.jobs_status = {}
+        self.jobs_terminated = {}
+        self.jobs_cleaned = {}
+        self.results = {}
+        self.errors = {}
+
+    def add_test_job(self, **kwargs):
+        """Create and track a db job object."""
+        job = job_factory(**kwargs)
+        self.jobs[job.id] = job
+        insert(job)
```

Comment:
What's this needed for?

Reply:
So, `get_reason_job_not_started` ends up querying the db for all running jobs.

I added it at one point to make things work, but then I changed how the db fixtures work, so it may not be needed any more. But there might be other places in the flow where the db is queryed for job data rather than using the provided `job` object.

Reply:
Ah, yes, and to check if a dependency is running or failed or what not.

Reply:
It seems surprising that it's in the TestAPI. I think that if the tests do need each job to be added to the DB then that should be done explicitly in the test, or in a helper function that saves it and adds it to the TestAPI.

Reply:
`add_test_job` is exactly that helper function?

Reply:
Ah, I think I see your point now - the factory for job create should do this, rather than the TestAPI. You're right - I'd come to this conclusion in parallel because we need SavedJobRequest state in the db to support the current manifest flow.

Reply:
Great.